File: SORTING/Bitonic_sort.py
import random
import time

"""*************Битонная сортировка - bitonic_sort************"""
def compare_swap(a, i, j, d):
    if (d == 1 and a[i] > a[j]) or (d == 0 and a[i] < a[j]):
        a[i], a[j] = a[j], a[i]

def merge(a, l, cnt, d):
    if cnt > 1:
        k = int(cnt / 2)
        for i in range(l, l + k):
            compare_swap(a, i, i + k, d)
        merge(a, l, k, d)
        merge(a, l + k, k, d)

def bitonic_sort(a, l, cnt, d):
    if cnt > 1:
        k = int(cnt / 2)
        bitonic_sort(a, l, k, 1)
        bitonic_sort(a, l + k, k, 0)
        merge(a, l, cnt, d)

def sort(a, N, u):
    bitonic_sort(a, 0, N, u)
"""***********************************************************"""

def create_array(size_element):
    arr = []
    for i in range(size_element):
        arr.append(round(random.random()*10000))
    return arr

def time_test(size):
    timer = time.time()
    my_array = create_array(size)
    timer = time.time() - timer
    print(f"Заполнение массива: {round(timer, 2)} сек")
    timer = time.time()
    sort(my_array, len(my_array), 1)
    timer = time.time() - timer
    print(f"Сортировка массива на {size} элементов: {round(timer,2)} сек")
    #print(my_array)

if __name__ == "__main__":

    time_test(10_000)

File: SORTING/Bitonic_sort.cpp
#include <iostream>

/*************Битонная сортировка - bitonic_sort************/
/*************Нестабильная версия***************************/
template <class T>
void compare(T* items, int i, int j, int dir) {
    if (dir == 1 && (items[i] > items[j]) || (dir == 0 && items[i] < items[j])) {
        std::swap(items[i], items[j]);
    }
}
// Sorts a bitonic sequence in the specified order.
template <class T>
void bitonic_merge(T* items, int lo, int n, bool dir) {
    if (n > 1) {
        int m = n / 2;
        for (int i = lo; i < lo + m; ++i) {
            compare(items, i, (i + m), dir);
        }
        bitonic_merge(items, lo, m, dir);
        bitonic_merge(items, lo + m, m, dir);
    }
}
// Sorts the given sequence in the specified order.
template <class T>
void bitonic_sort(T* items, int lo, int n, bool dir) {
    if (n > 1) {
        // Divide the array into two partitions and then sort 
        // the partitions in different directions.
        int m = n / 2;
        bitonic_sort(items, lo, m, 1);
        bitonic_sort(items, lo + m, m, 0);
        // Merge the results.
        bitonic_merge(items, lo, n, dir);
    }
}
// Sorts the given sequence in increasing order.
template <class T>
void bitonic_sort(T* items, int size) {
    bitonic_sort(items, 0, size, 1);
}
/***********************************************************/

void sort_test(int size_array) {
	setlocale(LC_ALL, "Russian");
	if (size_array < 1) { std::cout << "AAA";  return; }
	clock_t time = clock();
	int size = size_array;
	int* my_array = new int[size];
	std::cout << "Идёт заполнение массива: ";
	for (int i = 0; i < size; ++i) {
		my_array[i] = rand() % 1'000'000;
	}
	time = clock() - time;
	std::cout << double(time) / CLOCKS_PER_SEC << " сек." << std::endl;
	time = clock();
	std::cout << "Идёт сортировка массива на " << size << " элементов: ";
    bitonic_sort(my_array, size);
	time = clock() - time;
	std::cout << double(time) / CLOCKS_PER_SEC << " сек." << std::endl;
    /*
    for (int i = 0; i < size_array; ++i) {
        std::cout << my_array[i] << " ";
    }
    */
    delete[] my_array;
}

int main() {
    sort_test(100'000);
    return 0;
}
